### DISCUSSION TITLE
Autokey-gtk clipboard.get_selection() doesn't work.

### DISCUSSION THREAD

**Discussion by ineuw** [autokey/autokey#759] [Q&A] [answered]
This is definitely not caused by Autokey, not by Python and not by X-Windows . Nevertheless, its effect is that for the past two months **clipboard.get_ selection()** gets nothing. So, I did the following:

Uninstalled every keyboard and mouse utility, with the exception of **xsel**. 
Installed **Diodon** clipboard viewer, to show me the contents of of either sections **#CLIPBOARD** and/or the **#PRIMARY**.
(I am also using a wired keyboard and mouse )

**X Windows clipboard manager storage:**
**Control-c**, and the Context menu **copy** use the **#CLIPBOARD** 
**Control-a**, and **clipboard.get.selected()** use the **#PRIMARY** 
The mouse middle button also uses the  **#PRIMARY**, but haven't looked into this yet.

**Diodon** displays that Autokey, Python, function with X clipboard, but **clipboard.get_selection()** just ignores it. There are ~30 short scripts with the similar code structure, enclosing text with wiki codes and none work. But two long scripts using the same structure, do work.

My installs of **luckyBackup** and **Xfce4.note** also have problem with GTK.  Because of recent experiences involving GTK and/or Gnome, I decided to break down the relationship between X11, the keyboard, the mouse, and the interaction with the X clipboard. before considering changing platforms.

**Posted the following info in Pastebin:**
**Broken Autokey script example:**  https://pastebin.com/38m45wsn
**Working_textwrap_script:** https://pastebin.com/Bdu4VRRc
**Working_clear_text_script:** https://pastebin.com/Bdu4VRRc

**Inxi -Fxz System info:**  https://pastebin.com/9M2bDeBe
**List of user installed software:** https://pastebin.com/9M2bDeBe
**GTK libraries:** https://pastebin.com/p1kEctUA

**The standard error:**
```
Traceback (most recent call last):
  File "/home/ineuw/.local/lib/python3.10/site-packages/autokey/service.py", line 530, in _execute
    exec(compiled_code, scope)
  File "/home/ineuw/.config/autokey/data/english/superr-ht-ref-bil.py", line 19, in <module>
    text_=clipboard.get_selection()
  File "/home/ineuw/.local/lib/python3.10/site-packages/autokey/scripting/clipboard_gtk.py", line 93, in get_selection
    raise Exception("No text found in X selection")
Exception: No text found in X selection
```

**Comment by josephj11** [comment]
So far, I just looked at the first two scripts briefly. Comments:

Let's get past these first few issues (or misunderstandings on my part) and then see what's left to address.

First script:
Your try block can never fail (short of a catastrophic system error) because it doesn't "try" anything interesting.
It looks like your first API call should have been in there (right after the try statement) to catch it failing on an empty selection.

The try and except clauses appear to be reversed. When it succeeds (assuming the API call was in the try clause), it emits an empty reference. When it fails, it tries to emit the selection as a reference - which can't work because it failed to get a selection.

In your second script, there is no try block (because it just works). 

If I am reading man xsel correctly:
The first script starts  by clearing the primary selection, but the second script starts by clearing the clipboard. Why are these  different? They both seem to be using the primary selection at first.

The first script relies on the user making the selection and should emit an empty reference if they don't. The second script selects all the text automatically.

**Reply by ineuw** [reply]
Again a 1000 apologies for the late night errors. I focused on the explanations, and not on the script which I cobbled together for the post without testing. The original didn't have the test try: except: else:

The corrected script with the error is: https://pastebin.com/PYXb52Xa#QErrPuTP

**Comment by Elliria** [comment]
Interesting. It seems that the middle mouse button accesses **#PRIMARY** and **#SECONDARY**. I'm not sure that I'm able to correctly test **#SECONDARY** since I'm not sure any of my programs support it, but I messed around with both just now as well as I could.
- For **#PRIMARY**, I selected/highlighted some text and then pressed the middle mouse button.
- For **#SECONDARY**, I held down the alt or meta key while selecting/highlighting some text and then pressed the middle mouse button.

I'm finding that, in either case, the data that's pasted by the middle mouse button depends very much on the program that the selection was made in and on the program that it was pasted into. You can get interesting and sometimes inconsistent results by trying these tests in a variety of programs with either **#PRIMARY** or **#SECONDARY** selections or a mixture of both:

### Test 1
You may or may not get the same (original) selection each time instead of getting each of the selections when you try this:
1. Make a selection in **program A**.
2. Press the middle mouse button in **program A**.
3. Make another selection in **program A**.
4. Press the middle mouse button in **program A**.

### Test 2
This gets even more muddy if you jump between programs while making selections or pasting with the middle mouse button:
1. Make a selection in **program A**.
2. Press the middle mouse button in **program B**.
3. Make another selection in **program B**.
4. Press the middle mouse button in **program B**.
5. Press the middle mouse button in **program A**.

Even more interesting results can be obtained by testing in more than two programs at once.

**Reply by ineuw** [reply]
@Eliria, I have focused only on the keyboard, I rarely use the mouse. To highlight text for a script, I navigate with the keyboard keys. A primeval habit picked up before the dinosaurs' extinction, and never tried the middle button with the clipboard.

#SECONDARY is officially not-recommended, but probably used. Some offline apps interfere with copy and paste, and sometimes the results include text of the previous copy and paste.

**Reply by Elliria** [reply]
Yeah, I'm also a keyboard user, but am non-conventional (my favorite ways of getting and pasting text are **Shift+Delete** and **Shift+Insert** and some programs (and even Kubuntu) don't recognize those by default, causing some wrestling over here. I also never used the middle mouse button, but was curious about it for this particular use. Thanks for your description of offline apps interfering with copying and pasting. That explains the odd behavior I'm seeing.

**Comment by josephj11** [comment]
I think we can safely ignore the secondary selection. If it ever was in general use, that was very long ago.

If you need more clipboards, make your own with AutoKey global stored values. I made one and use it all the time.

Mine stores one string. Super+s is memory store. Super+r is memory recall and I have a memory clear script that I almost never use on some other hotkey. I use it when I'm doing cryptocurrency calculations using kcalc (which does not have a memory register) and I have to convert a number of values in their native token to their equivalent in US dollars so I need the regular clipboard for each of the values and my custom one for the conversion factor.

**Reply by ineuw** [reply]
I noticed that some apps supplement copy and paste of X. But, in my case, which concatenates segments of dynamically acquired data, global variables don't help. I use "xsel" to clear the X memory storage areas. 

I think you are missing my problem. I am asking you to test my script with Autokey-gtk if you have it. Failure or success will determine the next step. To me, the common culprit seems to be GTK. LMC is known to tinker and modify software.

All I need to know is the OS environment and compare it to Linux Mint Cinnamon. Try it on any ~100 character block of text and see if it surrounds the text with the <ref></ref> tags.

**Reply by josephj11** [reply]
The above comment was really for Elliria. I should have made that explicit.

**Reply by Elliria** [reply]
If the goal is to wrap the selected text with ref tags, I've got 15 selection wrapper scripts in AutoKey. Here's one of them that I just now tested on a 100-character block of text and it works for me:
```python
# This script replaces the selected string with a new string wrapped in the specified wrapper tags.
# Select some text in any window and click this script in the AutoKey context menu or press its hotkey.

# If a selection was made:
try:
    target = clipboard.get_selection()
    keyboard.send_keys("<ref>%s</ref>" % target)
# If no selection was made:
except:
    # Let the user know no selection was made:
    dialog.info_dialog(title="Error", message="No text was selected.")
```

**Reply by Elliria** [reply]
Although it's true that **#SECONDARY** is mostly extinct, it still exists and is still mentioned in web pages online, so it's worthy of noting, especially since we can access it from within AutoKey. It's testing it that's tricky, because it's tough to tell if we have software capable of it.

**Reply by Elliria** [reply]
Here's another of my 15 selection-wrapper scripts that toggles the wrapper, adding it if it's not there or removing it if it is:
```python
# This script replaces the selected string with a new string with the specified wrapper tags added if not present or removed if present.
# Select some text in any window and click this script in the AutoKey context menu or press its hotkey.

# If a selection was made:
try:
    # Store the selected text in a local variable:
    target = clipboard.get_selection()

    # Pause for a moment:
    time.sleep(0.2)

    # Check if the selected text contains wrapper tags:
    if "<ref>" not in target: 
        # Add wrapper tags if it doesn't:
        keyboard.send_keys("<ref>%s</ref>" % target)

    # Otherwise, if it has wrapper tags:
    else:
        # Delete the wrapper tags:
        keyboard.send_keys(target.replace("<ref>","").replace("</ref>",""))

# If no selection was made:      
except:
    # Let the user know no selection was made:
    dialog.info_dialog(title="Error", message="No text was selected.") 
```

**Reply by Elliria** [reply]
Last, but not least, here's yet another of my 15 selection-wrapper scripts that doesn't replace the selected text, but leaves it undisturbed while it wraps itself around it:
```python
# This scripts wraps text around the selection without manipulating the selection in any way.
# The script gets the contents of the clipboard (the selection) only to count its length.
# It then moves the cursor to the beginning of the selection in the focused window.
# It then inserts the specified opening wrapper text.
# It then moves the cursor to the end of the selection in the focused window.
# It then inserts the specified closing wrapper text.

# If a selection was made:
try:
    # Get the selection:
    text = clipboard.get_selection()
    # Get the length of the selection:
    length = len(text)
    # Move the cursor to the beginning of the selection:
    keyboard.send_keys("<left>")
    # Insert opening wrapper text:
    keyboard.send_keys("<ref>")
    # Move the cursor to the end of the selection:
    keyboard.send_key("<right>", repeat=length)
    # Insert closing wrapper text:
    keyboard.send_keys("</ref>")
# If no selection was made:
except:
    # Let the user know no selection was made:
    dialog.info_dialog(title="Error", message="No text was selected.")
```

**Comment by josephj11** [comment]
Solved! It was quite the puzzle.

This was getting weird. 

I changed the white space  in the code (because, although it was fine as is, it was hard for me to look at).

I removed all code that wasn't currently being used and ended up with this:

```
# [ super + r ] <ref></ref> tags to surround block of selected text
 
import os
##os.system("sleep .1; xsel -cp")  ## Don't do this!!
 
openref_ = "<ref>"
closeref_ = "</ref>"
 
try:
    time.sleep(0.1)
    text_ = clipboard.get_selection() ### fails
 
except:
    combined_ = (openref_ + closeref_)
 
else:
    time.sleep(0.1)
    text_ = text_.strip(" ")
 
    time.sleep(0.1)
    combined_ = (openref_ + text_ + closeref_)
 
time.sleep(0.1)
clipboard.fill_clipboard(combined_)
 
time.sleep(0.1)
keyboard.send_keys(combined_)
 
# os.system("sleep .1; xsel -cp")
 
```

I ran the original edited code: (all testing done in autokey.qt)

With no selection, it works as expected adding an empty reference.

With some text selected, if replaces the selected text with an empty reference - clearly wrong

With some text selected, but also copied to the clipboard, it works perfectly, replacing the selection with a reference containing the selection - This implies that get_selection() is working identically to get_clipboard() - maybe if the primary selection is empty but the clipboard is not, the clipboard becomes the primary selection or something like that.

Note: Your code (and the above code) places the result in the clipboard, but never uses it. The send_keys() produces the result.

Finally figured it out and now it works!

By selecting some text, you filled the primary selection, but the first thing your script does is call xsel which clears the primary selection. Don't do that! LOL

**Comment by ineuw** [comment]
Much much thanks to both of you. Not realizing that a selection is automatically identified by the PERMANENT clipboard, I defeated my efforts.

@Elliria, I loved both your solutions and the scripting style. Much thanks for that.

Here is the working script:

```
# [ super + r ] <ref></ref> tags to surround block of selected text
# 2023-01-18 04:11 - AutoKey

import os

selectall_ = "<ctrl>+a"
copy_ = "<ctrl>+c"
cut_ = "<ctrl>+x"
paste_ = "<ctrl>+v"

openref_ = "<ref>"
closeref_ = "</ref>"
left_ = "<left>"

try: 
    time.sleep(0.1)
    text_ = clipboard.get_selection()

except:
    time.sleep(0.1)
    combined_ = (openref_ + closeref_)

else:
    text_= text_.strip(" ")

    time.sleep(0.1)
    combined_= (openref_ + text_ + closeref_)

time.sleep(0.1)
keyboard.send_keys(combined_)

os.system("sleep .1; xsel -cp")
```

**Reply by josephj11** [reply]
Glad we fixed it. I am not aware of anything called a "permanent" clipboard.

You may want to take a look at our wiki [page](https://github.com/autokey/autokey/wiki/More-than-you-ever-wanted-to-know-about-X-clipboards) to see if anything should be added there to clarify the situation you just encountered.

**Reply by ineuw** [reply]
@josephj11, @Elliria, I found the webpage mentioned earlier. It's the clearest step by step explanation of the X-clipboard. Please help by adding it where you think it should be placed. [X selections ...](https://www.jwz.org/doc/x-cut-and-paste.html)

**Reply by josephj11** [reply]
Actually it was hiding in plain sight on the first line of our clipboard article. I added a couple of notes at the end of the existing article with another link to it and a few other tidbits.

**Reply by Elliria** [reply]
Thanks. That page also links to [a page on the Wayback Machine](https://web.archive.org/web/19980508141716/http%3A//tronche.com/gui/x/icccm/sec-2.html) that might be of interest.

**Reply by josephj11** [reply]
That page is interesting, but is written for application programmers and doesn't seem to add anything that an AutoKey user would need to know. I just skimmed it, so I may have missed something good...

**Reply by Elliria** [reply]
I figured @ineuw might find something useful in it.

**Comment by Elliria** [comment]
It looks great, @ineuw. I've added it to my collection. A girl can't have too many wrapper scripts.

**Reply by josephj11** [reply]
That's a wrap!

**Reply by ineuw** [reply]
My gratitude for contributing to my understanding. As for the errors, like PERMANENT instead of PRIMARY are due to inattention, brought on exhaustion, by the long hours and late nights seeking solutions to problems in the apps I use as tools.

**Reply by Elliria** [reply]
Oh, gosh, @josephj11, I should have seen that coming.

**Reply by Elliria** [reply]
Aw, here's a [virtual hug] for you, @ineuw. I hope your load lightens up soon.

**Reply by josephj11** [reply]
@ineuw Sometimes tool building seems to take more time than using them saves!

Tool building has been my favorite programming thing since I learned how to code. I'm always looking for ways to optimize my work flow and customize things to work "my way" - which is quite often different from what anybody else wants.

I pay for this highly when I have to do a fresh install of Linux. It usually takes me a couple of weeks to get everything back the way I want it.

I'm still trying to come up with  better way to remember all my changes. I get the most obvious ones and discover the rest when I try to do things and they don't work as expected.

**Reply by ineuw** [reply]
Thanks for the help. I am not too sure if I want the load to lighten. :-) It's a wonderful challenge re-learn what I forgot. I will revisit the help page, and add the link to the article that gave me the clearest explanation of the X clipboard and helped me to properly word this post.

**Reply by Elliria** [reply]
I use [Zim](https://zim-wiki.org/) for that, @josephj11 (it's in the Ubuntu repositories), with a page dedicated to the comprehensive steps for making all my usual changes to a fresh installation of GNU/Linux and then updating that as needed. I'd be lost without a personal wiki.

**Comment by josephj11** [comment]
@Elliria Just installed zim (again). I had it long ago, but didn't use it.

OFC it uses its own markup language.

**Reply by Elliria** [reply]
Yep, although you can use its interface as well, in which case it works like a word processor. My reason for choosing it was that MoinMoin didn't keep up with Python 3 [sigh] and I had to switch to something and this was, first of all, the only personal wiki I could find that didn't require running a server and truly is personal and, second of all, stores everything in plain text, so it's searchable with locate, grep, KRunner, and any other search programs. I've got an ever-growing AutoKey page in mine as well and won't be surprised to find that you will, too.

**Reply by josephj11** [reply]
I currently have a whole directory under $HOME/Public/howto/Autokey with 45 files in it. I don't think I'll import them all. I also have a hundred or more other files in folders at the same level as the Autokey folder. Importing all of that would take some work.

**Comment by Elliria** [comment]
I'll answer this in our Gitter conversation since it's going in a new direction from the original topic here.

**Comment by ineuw** [comment, marked as answer]
Please close this topic with my sincere thanks for your kind consideration and help. The problems are definitely unrelated to Autokey. I am composing a post for the Linux Mint forum, because I am confident that the developers actively scan the forum posts. Previously, I found a script installed in the .config folder testing LG monitor display problems. I will return later to post a link to the post.

**Reply by josephj11** [reply]
Just don't comment here any more. Discussions can be deleted, but I don't think they can be closed.
